src/main/java/org/opensearch/plugin/insights/core/service/categorizer/QueryShapeGenerator.java

```diff
@@ -195,11 +225,60 @@ static String buildSortShape(List<SortBuilder<?>> sortBuilderList, Boolean showF
      * @return String: comma separated list with leading space in square brackets
      * Ex: " [my_field, width:5]"
      */
-    static String buildFieldDataString(NamedWriteable builder) {
+    static String buildFieldDataString(NamedWriteable builder, Metadata metadata, Set<String> searchIndices) {
         List<String> fieldDataList = new ArrayList<>();
         if (builder instanceof WithFieldName) {
-            fieldDataList.add(((WithFieldName) builder).fieldName());
+            String fieldName = ((WithFieldName) builder).fieldName();
+            fieldDataList.add(fieldName);
+            String fieldType = getFieldType(fieldName, metadata, searchIndices);
+            if (fieldType != null) {
+                fieldDataList.add(fieldType);
+            }
         }
         return " [" + String.join(", ", fieldDataList) + "]";
     }
+
+    /**
+     * Method to get a field's type
+     * @return String field type
+     */
+    static String getFieldType(String fieldName, Metadata metadata, Set<String> searchIndices) {
+        if (metadata == null) {
+            return null;
+        }
+        for (String searchIndex : searchIndices) {
+            IndexMetadata indexMetadata = metadata.index(searchIndex);
+            if (indexMetadata == null) {
+                continue;
+            }
+            MappingMetadata mappingMetadata = indexMetadata.mapping();
+            if (mappingMetadata == null) {
+                continue;
+            }
+            Map<String, Object> sourceMap = mappingMetadata.getSourceAsMap();
+
+            String fieldType = findFieldType((Map<String, Object>) sourceMap.get("properties"), fieldName);
+            if (fieldType != null) {
```

Comment:
So we are just returning the first field type we find?
I thought we were planning to get field type for all indices and if the field type is the same then return it else consider it ambiguous and process it differently?

Reply:
My method is assuming a fieldName will be mapped to the same fieldType in all indices a large majority of the time. Also note, we are only checking the index mapping of indices which were searched as part of this request. 

We do not have any concrete data on this so I am open to either method. The current implementation saves some time and memory (very little).